### DISCUSSION TITLE
osu!taiko Classic+HD doesn't match stable

### DISCUSSION THREAD

**Discussion by zachmanthethird** [ppy/osu#17265] [Ideas]
Loosely related to #14794; off-handedly mentioned in https://github.com/ppy/osu/discussions/13047#discussioncomment-2312905
Screenshots provided courtesy of @Cookie-Tree as I'm now using Linux and cannot launch stable.
While the implementation of HD and its implications are still being discussed in #14798, I believe it makes sense for Classic+HD to replicate stable behavior instead of lazer behavior. While I don't think the "black box" confining the screen to 4:3 is necessary, I believe the 0% and 100% opacity points should be the same.
Pictured is https://osu.ppy.sh/b/3030622 at approximately 00:28:667 (notes prior were removed for the sake of accurate screenshots); lazer screenshots in version 2022.314.0
- Stable 16:9 
![image](https://user-images.githubusercontent.com/74262407/158475781-28393e51-777b-45e1-98e6-2f98b7794ee9.png)
- Lazer 16:9 including Classic
![image](https://user-images.githubusercontent.com/74262407/158475887-f21a4e04-dcb1-4c55-b158-1a6c280b90d4.png)
- Lazer 16:9 excluding Classic
![image](https://user-images.githubusercontent.com/74262407/158475918-47dc7d2f-4c56-49ce-ae6c-aa2d2af8eb88.png)

**Comment by bdach** [comment]
i do not know how to reimplement this without opening pandora's box of complexity. lazer mods were not designed to affect each other's operation like this, and i do not think they should be permitted to, either. if they are, then this is going to make pp calculation complexity and establishing mod incompatibility exponentially more difficult if mod authors abuse this sort of thing.

**Reply by zachmanthethird** [reply]
In that case, careful consideration will need to be taken when combining stable and lazer leaderboards. Not only will stable scores have to be recalculated (see https://github.com/ppy/osu/issues/14794#issuecomment-1059922788), but classic mod won't be comparative to stable scores. pp calculation for HD in lazer and HD in stable is already needing to be calculated separately, so I guess we just need to be extremely cautious as we continue.

**Reply by bdach** [reply]
my comment was aimed mostly at other developers/team members. a way to make this happen may be willed into existence depending on the team's position on this but i posit that allowing these sorts of mod interactions should not generally be allowed for non-legacy purposes.

**Reply by AadilRn** [reply]
how about just ignoring classic mod entirely and changing the HD mod to match the implementation of stable?

**Reply by frenzibyte** [reply]
I can imagine the hidden mod implementation being able to co-exist with "Classic" in a way where it would calculate the aspect ratio of the playfield in some form and add a cover to enforce the 4:3 visibility osu!(stable) provides (by using something like osu!mania's `PlayfieldCoveringWrapper`, one for covering the right-side of the playfield to enforce 4:3, and one for covering the left-side to match stable's "0% and 100% opacity points").

That however will have to depend on whether we want the playfield cover in hidden mod implementation, in which would appear at any aspect ratio because of how the playfield is locked to 16:9 without "Classic".

**Reply by bdach** [reply]
@frenzibyte note that this requires classic mod to be applied as late as possible so that it doesn't accidentally get applied *before* HD does. order enforcement adds a similar amount of complexity as allowing mixed mod effects does.

**Reply by frenzibyte** [reply]
The idea is that the aspect ratio would be calculated per-update (or using invalidations) to ensure that the order of mods doesn't matter (would also handle enabling/disabling mods in-game as well).

**Reply by bdach** [reply]
i mean that just sounds like swapping an explicit check for DT enabled for a very roundabout aspect ratio check. i dunno.

**Reply by frenzibyte** [reply]
I wouldn't exactly call it a swapping since it needs to know the aspect ratio to cover the rest of the playfield properly (which, without "Classic", would be enforced 16:9, and with "Classic", would be based on the game resolution), but we'll see how that goes.
